bundle/regal/rules/performance/non-loop-expression/non_loop_expression_test.rego

```diff
@@ -0,0 +1,389 @@
+				"row": 6,
+			},
+			"file": "policy.rego",
+			"row": 6,
+			"text": "\tendswith(input.email, \"acmecorp.com\")",
+		},
+		"related_resources": [{
+			"description": "documentation",
+			"ref": config.docs.resolve_url("$baseUrl/$category/non-loop-expression", "performance"),
+		}],
+		"title": "non-loop-expression",
+	}}
+}
+
+# this should be the basis for another rule, but for now it should be a non-error
+test_ok_unrelated_loops if {
+	r := rule.report with input as ast.policy(`
+allow if {
+	foo := input.foo[_]
+	input.foo == foo
+
+	some bar in data.bars
+	input.bar == bar
+}`)
+
+	r == set()
+}
+
+test_fail_two_somes if {
+	r := rule.report with input as ast.policy(`
+allow if {
+	some user in data.users
+	some email in user.emails
+	endswith(input.user_email, "acmecorp.com")
+	email == input.user_email
+}`)
+
+	r == {{
+		"category": "performance",
+		"description": "Non loop expression in loop",
+		"related_resources": [{
+			"description": "documentation",
+			"ref": config.docs.resolve_url("$baseUrl/$category/non-loop-expression", "performance"),
+		}],
+		"title": "non-loop-expression",
+		"location": {
+			"col": 2,
+			"file": "policy.rego",
+			"row": 7,
+			"end": {
+				"col": 44,
+				"row": 7,
+			},
+			"text": "\tendswith(input.user_email, \"acmecorp.com\")",
+		},
+		"level": "error",
+	}}
+}
+
+test_wildcard_assign if {
+	r := rule.report with input as ast.policy(`
+allow if {
+	user := data.users[_]
+	endswith(input.user_email, "acmecorp.com")
+	user.email == input.user_email
+}`)
+
+	r == {{
+		"category": "performance",
+		"description": "Non loop expression in loop",
+		"related_resources": [{
+			"description": "documentation",
+			"ref": config.docs.resolve_url("$baseUrl/$category/non-loop-expression", "performance"),
+		}],
+		"title": "non-loop-expression",
+		"location": {
+			"col": 2,
+			"file": "policy.rego",
+			"row": 6,
+			"end": {
+				"col": 44,
+				"row": 6,
+			},
+			"text": "\tendswith(input.user_email, \"acmecorp.com\")",
+		},
+		"level": "error",
+	}}
+}
+
+test_some_key_value if {
+	r := rule.report with input as ast.policy(`
+allow if {
+	some userID, user in data.users
+	endswith(input.user_email, "acmecorp.com")
+	user.email == input.user_email
+	userID != 0
+}`)
+
+	r == {{
+		"category": "performance",
+		"description": "Non loop expression in loop",
+		"related_resources": [{
+			"description": "documentation",
+			"ref": config.docs.resolve_url("$baseUrl/$category/non-loop-expression", "performance"),
+		}],
+		"title": "non-loop-expression",
+		"location": {
+			"col": 2,
+			"file": "policy.rego",
+			"row": 6,
+			"end": {
+				"col": 44,
+				"row": 6,
+			},
+			"text": "\tendswith(input.user_email, \"acmecorp.com\")",
+		},
+		"level": "error",
+	}}
+}
+
+test_every if {
+	r := rule.report with input as ast.policy(`
+allow if {
+    every role in data.required_roles {
+        endswith(input.email, "acmecorp.com")
+        role in input.roles
+    }
+}`)
+
+	r == {{
```

Comment:
We're repeating these large objects frequently here. Might want to look at how we use helpers in other tests to only pass the values that change between tests.

Reply:
I've used a with_location helper like we have in some other rules.